app.rb

```diff
@@ -22,6 +26,55 @@ def get_bank_code(code)
       end
       sublet_code || regular_code
     end
+
+    # Gets details of banks given the bank name,
+    # city and state
+    def filter_banks(state = nil, city = nil, bank = nil, limit = nil, offset = nil)
+
+      filtered_df = $df
+
+      unless state.nil?
+        filtered_df = filtered_df.where(filtered_df["STATE"].eq(state))
+      end
+
+      unless city.nil?
+        filtered_df = filtered_df.where(filtered_df["CITY"].eq(city))
```

Comment:
No support for fuzzy search?

Reply:
Right now the UI design doesn't require fuzzy search, hence we have implemented only regular search (user selects from a list of options). If we were to implement fuzzy search, for which the UI could be implemented later, what do you think would be the correct approach and the effort required to do so?

Reply:
No, experimenting with redisearch for the same. But trying to make sure that even if we switch the backend, we don't break the API.

Reply:
There's 35K unique values for `CITY`. And the breakdown by States isn't that helpful:

```
32,ANDAMAN AND NICOBAR ISLANDS
2694,ANDHRA PRADESH
65,ARUNACHAL PRADESH
628,ASSAM
1800,BIHAR
120,CHANDIGARH
519,CHHATTISGARH
26,DADRA AND NAGAR HAVELI AND DAMAN AND DIU
161,DELHI
181,GOA
1981,GUJARAT
881,HARYANA
704,HIMACHAL PRADESH
832,JAMMU AND KASHMIR
865,JHARKHAND
2309,KARNATAKA
1746,KERALA
13,LAKSHADWEEP
1502,MADHYA PRADESH
3182,MAHARASHTRA
67,MANIPUR
133,MEGHALAYA
31,MIZORAM
46,NAGALAND
1274,ODISHA
35,PUDUCHERRY
1417,PUNJAB
2094,RAJASTHAN
54,SIKKIM
2200,TAMIL NADU
813,TELANGANA
127,TRIPURA
3130,UTTAR PRADESH
477,UTTARAKHAND
2043,WEST BENGAL
```

So for someone searching in UP, WB, KA - the dropdown for cities will be 2000+.

Reply:
Right now, since we have implemented pagination, we are okay with the huge amount of results. We will discuss how these results will be managed in the UI side.

Reply:
@captn3m0 I just had a conversation with the product manager and I was told that, the flow of the UI will be such that the user selects the filters in this order

1. Bank name
2. State
3. City
4. Branch (I will be working on this filter right now)

Even if the intermediate results are huge, the pagination should be able to divide it. Once all 4 filters are applied, the number of results would be less.

Reply:
Switching from Bank Name to Bank Code might help a lot.

Reply:
Switched to bank code